### ISSUE TITLE
JsonFileCacheLayer - Use previous cache on next start

### ISSUE THREAD

**Issue by azhe403** [TurnerSoftware/CacheTower#199]
I use following snippet for registering `CacheStack` as `Singleton` into DI

```
services.AddSingleton(_ => {
    var stack = new CacheStack(new ICacheLayer[]
    {
        new JsonFileCacheLayer(cacheTowerPath)
    }, Array.Empty<ICacheExtension>());

    return stack;
});
```

For calling method I use `cacheStack.GetOrSetAsync<T>()` and they always refresh value instead of use previous cache (json file). 

It should can use previous JSON because I think cache isn't expired or staled, so refresh value is not need to be executed. 
Are any incorrect code or cannot injected it into DI?

### Environment (if applicable)

- NuGet Package Version: 0.11.16
- .NET Runtime Version: .NET 6
- Operating System: Windows

**Comment by Turnerj** [comment]
Hey @azhe403 - thanks for raising the issue! What you've described does sound odd and your code does look right here.
The file cache layer _should_ be pulling back the item from the cache assuming it isn't expired.

I'd love to know a little more about how your test case works when you're hitting this issue. Like, for example, is your call to `GetOrSetAsync` done and a few moments later, you call it again to see it not set? Is it within the same application instance? Is this after you restart the application?

Could you also provide the arguments you're using for `GetOrSetAsync`? Specifically, what you're setting for `CacheSettings`.

**Comment by azhe403** [comment]
I use following snippet code.
```
var cache = await _cacheStack.GetOrSetAsync<T>(cacheKey, async (_) => await action(), 
_cacheSettings = new CacheSettings(TimeSpan.FromMinutes(_expireAfter), TimeSpan.FromMinutes(_staleAfter)));
```

where `_expireAfter `is 60, and `_staleAfter` is 10

If you want to see my ugly code, you can find here
https://github.com/WinTenDev/ZiziBot.NET/blob/df6cd808ea528f03033014edda63fa92402e4836/src/WinTenDev.Zizi.Services/Internals/CacheService.cs#L85
https://github.com/WinTenDev/ZiziBot.NET/blob/df6cd808ea528f03033014edda63fa92402e4836/src/WinTenDev.Zizi.Utils/Extensions/CacheTowerServiceExtension.cs#L14

Btw I'm use **CacheTower** for **ASP.NET** application, so it will be create new instance every restart.

**Comment by Turnerj** [comment]
Can you try:
- Starting your application
- Calling the code path that invokes the cache
- Stopping your application
- Checking the manifest file for the file cache (this is stored in the path you supply to Cache Tower in your DI setup)

I'm expecting to see an entry in the manifest file that should include the cache key and when it expires.

Additionally after doing that, repeat the exact same process and share _that_ manifest file too.

Basically what you describe sounds like the manifest file isn't saving. It is designed to only save the manifest file on cache stack dispose which _should_ happen when the application stops (unless you have something like an Out Of Memory crash). [I am contemplating having manifest files save at regular intervals though.](https://github.com/TurnerSoftware/CacheTower/issues/187#issuecomment-907779681)

**Comment by azhe403** [comment]
Where to specify interval for saving to `manifest.json` because some time `manifest.json` file not be updated (file still empty value)

or it can be manual trigger by method?

**Comment by Turnerj** [comment]
> Where to specify interval for saving to `manifest.json` because some time `manifest.json` file not be updated (file still empty value)
> 
> or it can be manual trigger by method?

Sorry, I meant in my other comment that I'm contemplating adding it - like in the future, I might add it just in case. It _shouldn't_ be necessary as it is intended to save the manifest on dispose of the Cache Stack which would happen when the application closes. A full application or system crash would prevent the manifest from saving but those should be rare (plus losing cached values is less important comparatively).

**Comment by pablopioli** [comment]
I had manifest corruption problems in some environments where the applicacion crashed randomly. In those cases all cache is busted because you get an exception just reading values. Just to let you know that this can be a worse problem than losing cached values.

In my case I solved it writing my own cache layer using a brute force approach. 
https://github.com/pablopioli/FileCache
Hope it helps someone.

**Comment by Turnerj** [comment]
Hey @pablopioli - do you have an example of a corrupt manifest? Do you mean like the JSON is invalid in it? Or maybe the manifest listed a file that no longer exists?

This is the first I'm hearing about exceptions when retrieving an item from the cache from a corrupt manifest. In other cases with issues to the manifest, it just results in the cache item being re-requested, not an exception.

**Comment by pablopioli** [comment]
In one case I found something like this

`{"Value":{}}enantS1":{"FileName":"294D8C273914589CAE5487312B86C122.json","Expiry":"2022-01-29T14:41:59Z"}}}`

Obviously that's more a file system problem than a problem with the library, I couldn't pin point the exact cause because it's attached storage to a Kubernetes pod.

My main problem would go like this. You evict an item from the cache, but the manifest didn't got saved.

The next time you got an exception 
`Could not find file '/folder/294D8C273914589CAE5487312B86C122.json'`
because the manifest is not in sync with the physical files.

I see your approach of using a manifest is more efficient but if your environment is unstable you can get into trouble soon.

**Comment by aljones** [comment]
The corruption might be caused by the use of [FileMode.OpenOrCreate](https://github.com/TurnerSoftware/CacheTower/blob/082e9cd2b327c7b42d34068081f9105d31337703/src/CacheTower/Providers/FileSystem/FileCacheLayerBase.cs#L75). It should probably be be FileMode.Create, which will truncate an existing file.

**Comment by Turnerj** [comment]
That's a good point @aljones - that actually could be `FileMode.Create` and would be more appropriate because we are only writing, not reading.

I'm hoping to have time in the next few weeks to try and push through my changes for serializers (#172) where I'll likely tack that `FileMode` change in.

Regarding the main point of this issue though, I'm thinking of two solutions:
- Dropping manifests and using file system APIs (eg. files in directory) to provide "manifest" support
- Having manifests periodically save

Dropping manifests remove one point of failure and just rely on the file system access of a file to say it exists or not. The problem with this is that it makes cleanup tasks more expensive as each cache entry file needs to be read for an expiry. Reading each file means deserializing each file to find the expiry date which, depending on the data cached, could be _very_ expensive for both processing and allocations.

Having manifests periodically save would avoid the issue of crashing and no manifest being saved. It doesn't prevent _corrupt_ manifests but that may be partially addressed by the suggested `FileMode` change.

I'm leaning towards just periodically saving manifests. I'd probably also add a secondary check of `File.Exists` before attempting to read the file too to avoid that exception @pablopioli raised - I would rather return a cache miss here than throw an exception.
